system_zero/src/alu/bitops.rs

```diff
@@ -0,0 +1,473 @@
+use plonky2::field::extension_field::Extendable;
+use plonky2::field::field_types::{Field, PrimeField64};
+use plonky2::field::packed_field::PackedField;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use starky::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+
+use crate::registers::alu::*;
+use crate::registers::NUM_COLUMNS;
+
+/// Interpret the first 32 elements of `bits` as bits from low to high
+/// of a u32 and return \sum_i bits[i] 2^i as an element of P.
+///
+/// TODO: This should probably raise an error if there are more than
+/// 32 elements in `bits`.
+fn binary_to_u32<F, P, I>(bits: I) -> P
```

Comment:
I would suggest defining `I` as `&[P; 32]`, so the type system enforces the length.